### PR TITLE
Media: Ensure user feedback when inserting unsupported file types via WordPress media modal

### LINKED ISSUE #6756: Media: No feedback when inserting unsupported file types via WordPress media modal
## Bug Description

Sometimes when users try to add a video to a story page by using the Upload button and press "Insert into Page" it doesn't actually do anything

## Expected Behaviour

After the upload completes, and the user presses the "Insert into page" button the video asset should be place on the currently selected page

## Steps to Reproduce

1. In the editor, select the media upload tab
2. Select upload media
3. Upload a video asset via the WP Media upload modal
4. Select "Insert into Story" from the modal when upload is completed.
5. Feel the warmth of your tears as they fall down your face in disappointment.

https://user-images.githubusercontent.com/58738530/110989073-5bb99780-833f-11eb-91bf-f2366fd3c963.mov

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

This ticket become a little larger, when I looked around a different uses of the media picker in editor. In short, the link ( elements) , poster (document), publisher logo ( document ), caption ( video ) and poster ( video ), did have any feedback / error handling if an incorrect file type was selected. Now a snack message is displayed in all these different cases with an error message that matches what you are doing. 

## Relevant Technical Choices

- New PHP changes are required, so that a list of image types can be displayed. 
- Link icons, now only support valid image types and ignore SVGS. 
- Captions are now limited to VTT files and show an error message when an invalid one is selected.

## To-do

- Add some tests
- Name variable and param names better. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

Test 1: Insert in compatibility file types for media dialog. 

1. Ensure that transcoding feature flag is disabled.
2. Open the editor. 
3. Click upload button. 
4. Upload an incompatible file type, such as a .mov or .pdf file. 
5. Wait for upload to complete.
6. Click insert into page. 
7. See showbar message.

Test 2: Select poster / publish logo

0. Open the editor. 
0. Click to open documents tab. 
0. Click on edit icon for poster or publisher logo
0. Upload an incompatible file type, such as a .mov or .pdf file. 
0. Wait for upload to complete.
0. Click select poster / publisher logo button
0. See showbar message.

Test 3: Select poster for video element

0. Open the editor. 
0. Insert video into canvas ( either by uploading a new one or using an existing one, not important for this test ).
0. Click on edit icon for poster
0. Upload an incompatible file type, such as a .mov or .pdf file. 
0. Wait for upload to complete.
0. Click select poster
0. See showbar message.

Test 4: Select caption file for video element

0. Open the editor. 
0. Insert video into canvas ( either by uploading a new one or using an existing one, not important for this test ).
0. Click on upload captions button
0. Upload an incompatible file type, such as a .mov or .pdf file. 
0. Wait for upload to complete.
0. Click select poster
0. See showbar message.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6756

### CODE REVIEW

File: assets/src/edit-story/components/mediaPicker/useMediaPicker.js [hunk 1]
```diff
@@ -44,6 +46,7 @@ export default function useMediaPicker({
   title = __('Upload to Story', 'web-stories'),
   buttonInsertText = __('Insert into page', 'web-stories'),
   onSelect,
+  onSelectErrorMessage = __('Unable to use this file type. ', 'web-stories'),
```
Comment (swissspidy) [inline assets/src/edit-story/components/mediaPicker/useMediaPicker.js]:
```suggestion
  onSelectErrorMessage = __('Unable to use this file type.', 'web-stories'),
```

File: assets/src/edit-story/components/panels/design/link/link.js [hunk 2]
```diff
@@ -166,6 +168,18 @@ function LinkPanel({ selectedElements, pushUpdateForObject }) {
     [handleChange]
   );
 
+  const iconErrorMessage = useMemo(() => {
+    /* translators: %s is a list of allowed file extensions. */
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/design/link/link.js]:
```suggestion
```

File: assets/src/edit-story/components/panels/design/poster/videoPoster.js [hunk 3]
```diff
@@ -54,12 +54,25 @@ function VideoPosterPanel({ selectedElements, pushUpdate }) {
     [pushUpdate, rawPoster]
   );
 
+  const posterErrorMessage = useMemo(() => {
+    /* translators: %s is a list of allowed file extensions. */
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/design/poster/videoPoster.js]:
```suggestion
```

File: assets/src/edit-story/components/panels/document/publish/publish.js [hunk 4]
```diff
@@ -165,6 +169,30 @@ function PublishPanel() {
     [updateStory]
   );
 
+  const publisherLogoErrorMessage = useMemo(() => {
+    /* translators: %s is a list of allowed file extensions. */
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/document/publish/publish.js]:
```suggestion
```

File: assets/src/edit-story/components/panels/document/publish/publish.js [hunk 5]
```diff
@@ -165,6 +169,30 @@ function PublishPanel() {
...
+      __('Please choose only %s as publisher logo.', 'web-stories'),
+      allowedImageFileTypes.join(
+        /* translators: delimiter used in a list */
+        __(', ', 'web-stories')
+      )
+    );
+  }, [allowedImageFileTypes]);
+
+  const posterErrorMessage = useMemo(() => {
+    /* translators: %s is a list of allowed file extensions. */
```
Comment (swissspidy) [inline assets/src/edit-story/components/panels/document/publish/publish.js]:
```suggestion
```

File: assets/src/edit-story/components/library/panes/media/local/mediaPane.js [hunk 6]
```diff
@@ -197,8 +182,19 @@ function MediaPane(props) {
     }
   };
 
+  /* translators: %s is a list of allowed file extensions. */
```
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/local/mediaPane.js]:
```suggestion
```

**Review by swissspidy — APPROVED** [review]
LGTM 👍

**Comment by swissspidy** [comment]
Still needs QA instructions btw

**Comment by spacedmonkey** [comment]
QA details added. 

@csossi Just a note for the future, the 4 tests I listed above are all the places currently that the media picker is used in the editor. Any we make tweaks to the behaviour of the media picker, please keep these tests in mind, as you may find bugs.

**Comment by swissspidy** [comment]
@spacedmonkey You'll need to add `allowedImageMimeTypes` and `allowedImageFileTypes` here to fix Karma tests:

https://github.com/google/web-stories-wp/blob/5ac7794a11a2aa8734b2b573540f7afcd5638b08/assets/src/edit-story/karma/fixture/fixture.js#L43-L65

**Comment by swissspidy** [comment]
A karma test in `videoPoster.karma.js` seems to be failing:

https://github.com/google/web-stories-wp/pull/6772/checks?check_run_id=2142657332#step:9:3330

**Comment by spacedmonkey** [comment]
> A karma test in videoPoster.karma.js seems to be failing:

I am struggling to work out what this one is. I wonder if it is unrelated. [This](https://github.com/google/web-stories-wp/blob/d872f46a817b094777ceb07109be5a234effbbd6/assets/src/edit-story/components/panels/design/poster/karma/videoPoster.karma.js#L58-L82) is the test that is failing. What is interesting is that reset via mouse that is failing, but in https://github.com/google/web-stories-wp/commit/587779e44822d5c60e96e00259c5ff68b9d28d6f @miina skipped the keyboard version of this test. I wonder if the mouse version of the test is now flakey as well. 

@miina Can you give some more context on why this test was skipped?

**Comment by swissspidy** [comment]
Hmm `useMediaPiicker` displays the error message on `if (Array.isArray(type) && !type.includes(mediaPickerEl.mime)) {`

In the Karma test only `url` is mocked, but `mime` is missing:

https://github.com/google/web-stories-wp/blob/d872f46a817b094777ceb07109be5a234effbbd6/assets/src/edit-story/components/panels/design/poster/karma/videoPoster.karma.js#L47

Maybe it's because of that?

Could add it like this I suppose:

```js
window.wp = {
  ...window.wp,
  media: () => ({
    state: () => ({
      get: () => ({
        first: () => ({
          toJSON: () => ({
            url: 'http://example.com/dummy.jpg',
            mime: 'image/jpeg',
          }),
        }),
      }),
    }),
    on: (type, callback) => callback(),
    once: (type, callback) => callback(),
    open: () => {},
  }),
};
```

**Comment by miina** [comment]
> @miina Can you give some more context on why this test was skipped?

Just because it was flaky and was stopping an unrelated PR getting merged by needing to rerun the Karma tests. It was occasionally failing in other PRs, too. Not sure about the mouse one being flaky -- that I haven't seen failing as far as I remember.

**Comment by spacedmonkey** [comment]
> Hmm useMediaPiicker displays the error message on if (Array.isArray(type) && !type.includes(mediaPickerEl.mime))

Thanks for this find. I haven't done a lot of work on karma tests, so they were new to me. Thanks for the support.

**Pull request merged**